File: code/java/projects/huffman/alg/Huffman.java
package alg;

import java.util.Comparator;
import java.util.Iterator;

import adt.Bag;
import adt.Set;
import adt.PriorityQueue;
import impl.BitSequence;
import impl.HeapPriorityQueue;
import impl.MapBag;
import impl.MapSet;


/**
 * Class to implement the Huffman encoding.
 * 
 * @author Thomas VanDrunen
 * Algorithmic Commonplaces
 * May 18, 2020
 */
public class Huffman {

    // ---- nested classes supporting the encoding ---
    
    /**
     * Supertype to represent nodes in a Huffman tree
     */
    private static abstract class HuffmanNode {
        /** The characters represented in the tree rooted here */
        char[] letters;
        /** The combined frequencies of all the characters in this subtree */
        int freq;

        /** 
         * Does this subtree contain the given character?
         * @param c The character to search for
         * @return True if this subtree contains the given character, false otherwise
         */
        boolean contains(char c) {
            // bounded linear search on the letters array
            boolean found = false;
            for (int i = 0; ! found && i < letters.length; i++)
                found = letters[i] == c;
            return found;
        }

        /**
         * Encode a given character by adding appropriate bits to a
         * given bit sequence.
         * Precondition: contains(c)
         * Postcondition: Bits have been added to bSeq indicating the path
         * in the tree rooted at this node to the leaf representing the given
         * character.
         * @param c The character to encode
         * @param bSeq The bits encoding the given character
         */
        abstract void encode(char c, BitSequence bSeq);

        /**
         * Find the character that results in a search from this node
         * as directed by the given boolean iterator.
         * Postcondition: The prefix of bits that leads to a leaf in the
         * tree has been expended by the iterator.
         * @param it The iterator directing the search
         * @return The character found
         */
        abstract char decode(Iterator<Boolean> it);

        /** 
         * Retrieve the combined frequencies of all the characters 
         * in this subtree 
         * @return The combined frequency.
         */
        int getFreq() { return freq; }
    }

    /**
     * Class to represent a leaf in a Huffman tree
     */
    private static class Leaf extends HuffmanNode {

        /** Constructor, taking a character and its frequency */
        Leaf(char c, int freq) { 
            letters = new char[1];
            letters[0] = c;
            this.freq = freq;
        }
        
        /** 
         * Encode a given character by adding appropriate bits to a
         * given bit sequence.
         * Precondition: This leaf represents the given character
         * Postcondition: No bits (that is, an empty bit sequence) has been added
         * to the given bitsequence
         */
        public void encode(char c, BitSequence bSeq) {
            assert letters[0] == c;
        }

        /**
         * Find the character that results in a search from this node
         * as directed by the given boolean iterator.
         * Postcondition: No bits have been expended by the iterator.
         * @param it The iterator directing the search
         * @return The character represented here
         */
        public char decode(Iterator<Boolean> it) { return letters[0]; }
    }

    /**
     * Class representing an internal node in a Huffman tree.
     *
     */
    private static class Internal extends HuffmanNode {
        /** The children of this node */
        HuffmanNode left, right;

        /**
         * Constructor taking the two children
         */
        Internal(HuffmanNode left, HuffmanNode right) {
            this.left = left;
            this.right = right;
            letters = new char[left.letters.length + right.letters.length];
            int i = 0;
            for (int j = 0; j < left.letters.length; j++)
                letters[i++] = left.letters[j];
            for (int j = 0; j < right.letters.length; j++)
                letters[i++] = right.letters[j];
        }
        
        /**
         * Encode a given character by adding appropriate bits to a
         * given bit sequence.
         */
        public void encode(char c, BitSequence bSeq) {
             throw new UnsupportedOperationException();
        }

        /**
         * Find the character that results in a search from this node
         * as directed by the given boolean iterator.
         */
       public char decode(Iterator<Boolean> it) {
           // This call consumes exactly one value from the iterator
           assert it.hasNext();
           return (it.next()? left : right).decode(it);
        }
    }
    
    /**
     * Class to represent encoded messages.
     * No public interface---this is returned from encode()
     * and can be passed to decode().
     */
    public static class EncodedMessage {
        private EncodedMessage(HuffmanNode root, BitSequence bits) {
            this.root = root;
            this.bits = bits;
        }
        private final HuffmanNode root;
        private final BitSequence bits;
    }
    
    /**
     * Encode a message
     * @param msg The plaintext to be encoded.
     * @return An object represented an encoded version of the text
     */
    public static EncodedMessage encode(String msg) {
        // Analyze the characters in msg by frequency
        Bag<Character> freqs = new MapBag<Character>();
        Set<Character> chars = new MapSet<Character>();
        for (char c : msg.toCharArray()) {
            freqs.add(c);
            chars.add(c);
        }
        int numChars = chars.size();

        // Initialize the worklist as a priority queue
        PriorityQueue<HuffmanNode> worklist = new HeapPriorityQueue<HuffmanNode>(
                numChars, 
                 null // need a Comparator
                );

        // The minimal prefix-code key for this message
        HuffmanNode key;

        // --- Compute minimal key ---
         // ADD CODE HERE

        // Encode the message using the key
        BitSequence bits = new BitSequence();
        for (char c : msg.toCharArray())
            key.encode(c, bits);
        return new EncodedMessage(key, bits);
    }
    
    /**
     * Decode a message.
     * @param msg An object standing for an encoded message
     * @return The plaintext decoded message.
     */
    public static String decode(EncodedMessage msg) {
        String decodedMsg = "";
        for (Iterator<Boolean> it = msg.bits.iterator(); it.hasNext();)
            decodedMsg += msg.root.decode(it);
        return decodedMsg;
    }
    
}

File: code/python/huffman/huffman.py
'''
Created on Jan 14, 2020

@author: tvandrun
'''

from bag import DictBag
from priorityqueue import PriorityQueue
from functools import reduce

# Class to represent a leaf node in a Huffman key
class Leaf :    
    def __init__(self, c):
        # The character represented by this leaf
        self.c = c
        # The set of characters in the subtree rooted here,
        # which is just the one letter
        self.letters = set([c])
    def contains(self, x):
        return x == self.c
    def encode(self, x):
        return []
    def decode(self, bits):
        return (self.c, bits)
    def __str__(self):
        return "{" + self.c + "}"

# Class to represent an internal node in a Huffman key
class Internal :
    def __init__(self, left, right):
        # The left and right children
        self.left = left
        self.right = right
        # The set of characters in the subtree rooted here
        self.letters = left.letters.union(right.letters)
    def contains(self, x):
        return (x in self.letters)
    def encode(self, x):
        if self.left.contains(x) :
            return [True] + self.left.encode(x)
        else :
            #print(x)
            assert self.right.contains(x) 
            return [False] + self.right.encode(x)
    def decode(self, bits):
        if bits[0] :
            return self.left.decode(bits[1:])
        else :
            return self.right.decode(bits[1:])
    def __str__(self):
        return "["+str(self.left) + "," + str(self.right) + "]";

def char_encoding_to_bit_string(c, key):
    to_return = ""
    for b in key.encode(c) :
        to_return += "1" if b else "0"
    return to_return

def msg_encoding_to_bit_string(msg, key):
    to_return = ""
    for c in msg:
        to_return += char_encoding_to_bit_string(c, key)
    return to_return

def encode(msg):
    # Analyze the characters in msg by frequency
    frequencies = DictBag()
    characters = set([])
    for c in msg :
        frequencies.add(c)
        characters.add(c)

    # Initialize the worklist to be a priority queue of
    # node, combined-frequency pairs, with a lower combined
    # frequency as a higher priority
    worklist = PriorityQueue(len(characters), lambda x, y : x[1] < y[1])

    # Make leaves for all the characters, inserting them in the worklist
    for c in characters :
        worklist.insert((Leaf(c), frequencies[c]))

    # Repeatedly remove the two nodes with least combined frequency,
    # unite them, and insert the new node into the worklist
    for i in range(len(characters)-1):
        (left, left_freq) = worklist.extract_max()
        (right, right_freq) = worklist.extract_max()
        merged = Internal(left, right)
        worklist.insert((merged, left_freq + right_freq))

    # Retrieve the remaining node, which is the root of the final key
    key = worklist.extract_max()[0]
    assert worklist.is_empty()
    for c in characters :
        print(str(c) + "&" + char_encoding_to_bit_string(c, key))

    # Encode the message using the key as key
    encoding = reduce(lambda bits, c : bits + key.encode(c), msg, [])
    ## Without using reduce:
    #encoding = []
    #for c in msg :
    #    encoding += key.encode(c)

    # Return the encoded message and key
    return (encoding, key)

def encode_from_tree(msg, key):
    bitseq = []
    for c in msg:
        bitseq += key.encode(c)
    return bitseq
    
def encode_from_tree_redux(msg, key):
    return reduce(lambda bits, c : bits + key.encode(c), msg, [])

def decode(msg, key):
    decoded_msg = ""
    print(str(len(msg)))
    while len(msg) > 0 :
        (c, msg) = key.decode(msg)
        decoded_msg += c
    return decoded_msg

(encoded_msg, key) = encode("IF THIS IS A CONSULAR SHIP THEN WHERE IS THE AMBASSADOR")
print(msg_encoding_to_bit_string("COMPLETE", key))

import unittest

class HuffmanTest(unittest.TestCase):
    def checkMsg(self, msg):
        (encoded_msg, key) = encode(msg)
        self.assertEqual(msg, decode(encoded_msg, key))
    
    def testVader(self):
        self.checkMsg("IF THIS IS A CONSULAR SHIP THEN WHERE IS THE AMBASSADOR")
        